modules/extensions-mule-support/src/main/java/org/mule/runtime/module/extension/mule/internal/loader/MuleSdkPluginExtensionModelLoader.java

```diff
@@ -0,0 +1,111 @@
+
+/**
+ * {@link ExtensionModelLoader} implementation for Mule SDK Plugins.
+ *
+ * @since 4.5.0
+ */
+public class MuleSdkPluginExtensionModelLoader extends AbstractExtensionModelLoader {
+
+  private static final Logger LOGGER = getLogger(MuleSdkPluginExtensionModelParserFactory.class);
+
+  @Override
+  public String getId() {
+    return MULE_SDK_EXTENSION_LOADER_ID;
+  }
+
+  @Override
+  protected ExtensionModelParserFactory getExtensionModelParserFactory(ExtensionLoadingContext context) {
+    return new MuleSdkPluginExtensionModelParserFactory();
+  }
+
+  @Override
+  public ExtensionModel loadExtensionModel(ExtensionModelLoadingRequest request) {
+    Optional<ExtensionModel> extensionModel = Optional.empty();
+
+    // Parses the AST if not already available in the loading context.
+    ArtifactAst ast = (ArtifactAst) request.getParameters().get(MULE_SDK_ARTIFACT_AST_PROPERTY_NAME);
+    if (ast == null) {
+      // Note that during parsing of the AST, the ExtensionModel will need to be generated in order to perform validations.
+      extensionModel = parseAstAndBuildPluginExtensionModel(request);
+    }
+
+    // If the ExtensionModel has already been generated, return that, otherwise, do the actual loading now.
+    return extensionModel.orElseGet(() -> super.loadExtensionModel(request));
+  }
+
+  private AstXmlParser createAstParser(Set<ExtensionModel> dependencies, boolean disableValidations) {
+    AstXmlParser.Builder astBuilder = builder()
+        .withArtifactType(MULE_EXTENSION)
+        .withExtensionModels(dependencies);
+
+    if (disableValidations) {
+      astBuilder.withSchemaValidationsDisabled();
+    }
+
+    return astBuilder.build();
+  }
+
+  private Optional<ExtensionModel> parseAstAndBuildPluginExtensionModel(ExtensionModelLoadingRequest request) {
+    Set<ExtensionModel> dependencies = request.getDslResolvingContext().getExtensions();
+
+    String[] resources = {getRequiredLoadingParameter(request, MULE_SDK_RESOURCE_PROPERTY_NAME)};
+    MuleSdkExtensionModelLoadingHelper loadingHelper =
+        new MuleSdkPluginExtensionModelLoadingHelper(request.getArtifactCoordinates());
+
+    try {
+      // Parses the full AST of the artifact by providing a helper for loading the ExtensionModel that represents the artifact
+      // itself, this is so that schema validations can be performed properly.
+      ArtifactAst artifactAst = parseArtifact(resources,
+                                              this::createAstParser,
+                                              dependencies,
+                                              false,
+                                              request.getExtensionClassLoader(),
+                                              loadingHelper);
+
+      // Applies the AST validators and throws if there was any error
+      logWarningsAndThrowIfContainsErrors(validatorBuilder().build().validate(artifactAst), LOGGER);
+      return loadingHelper.getExtensionModel();
```

Comment:
won't this end up looking up for another instance of this same loader in order to create the extension model?

Reply:
The `getExtensionModel` method wouldn't, the `loadExtensionModel` would. That one gets called as part of the parsing. The `getExtensionModel` method doesn't exist anymore now.

The `loadExtensionModel` still works as it was. We don't need it to be the same instance of the loader for it to work (because it is stateless), but we could change it to use the same instance if we think it makes more sense, or if it we just want to avoid creating another instance (with the associated enrichers and validators) for performance reasons.

Reply:
Done, reusing the same loader instance now.